### DISCUSSION TITLE
ref struct and return ref this.

### DISCUSSION THREAD

**Discussion by lieene** [dotnet/csharplang#4046] [General] [answered]
**ref struct** is supposed to live only on stack and unmanaged.
in a struct like
```csharp
ref struct CanReturnRef
{
    public CanReturnRef( int i ) => this.i=i;
    private int i;
    public int Value=> i;
    public ref CanReturnRef Next()
    {
        i++;
        return ref this;
    }
}
```
`return ref this;` Should be allowed.
So a Fluent design will work better;
for example
```csharp
{
    Print( new CanReturnRef(0).Next().Next().Next().Value );//output 3
    var fluent = new CanReturnRef(0);
    fluent.Next().Next().Next();
    Print(fluent.Value);//output 1 which is not expected, but can be fixed by return ref this.
}
```

**Comment by canton7** [comment]
See also: https://github.com/dotnet/csharplang/discussions/1997

**Reply by lieene** [reply]
#1997 is for general struct where the case is much more complicated. 
A ref struct can not be used as a field of any managed type.
So return ref this does not involve any GC pin problem. It's simple and valid.

**Comment by svick** [comment, marked as answer]
[There is already a proposal that will allow you to do this](https://github.com/dotnet/csharplang/blob/master/proposals/low-level-struct-improvements.md#provide-struct-this-escape-annotation). As far as I can tell, the championed issue for that is https://github.com/dotnet/csharplang/issues/1147.
